File: app/src/main/java/com/nizarmah/sada/viewmodel/SmsViewModel.kt
package com.nizarmah.sada.viewmodel

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.net.HttpURLConnection
import java.net.URL
import com.google.gson.Gson

import com.nizarmah.sada.model.Message
import com.nizarmah.sada.util.NetworkUtils
import com.nizarmah.sada.util.PermissionsHelper
import com.nizarmah.sada.util.PermissionsManager
import java.util.UUID

private const val TOWER_PORT = 8080

class SmsViewModel(app: Application) : AndroidViewModel(app) {

    private val gson = Gson()

    val permissions = PermissionsHelper.getSmsPermissions()
    val permissionsGranted = PermissionsManager.smsPermitted

    private val _deviceIp = MutableStateFlow<String>("")
    val deviceIp: StateFlow<String> = _deviceIp

    private val _towerIp = MutableStateFlow<String>("")
    val towerIp: StateFlow<String> = _towerIp

    private val _towerOnline = MutableStateFlow<Boolean>(false)
    val towerOnline: StateFlow<Boolean> = _towerOnline

    private val _sendError = MutableStateFlow(false)
    val sendError: StateFlow<Boolean> = _sendError.asStateFlow()

    private val _messageText = MutableStateFlow("")
    val messageText: StateFlow<String> = _messageText.asStateFlow()

    private val _username = MutableStateFlow("")
    val username: StateFlow<String> = _username.asStateFlow()

    private val _messages = MutableStateFlow<List<Message>>(emptyList())
    val messages: StateFlow<List<Message>> = _messages.asStateFlow()

    init {
        viewModelScope.launch {
            _deviceIp.value = NetworkUtils.getLocalIpAddress().orEmpty()
            _towerIp.value = NetworkUtils.getGatewayIpAddress(app).orEmpty()
            _towerOnline.value = isTowerOnline()
        }
    }

    private suspend fun isTowerOnline(): Boolean = withContext(Dispatchers.IO) {
        if (_towerIp.value.isEmpty()) {
            return@withContext false
        }

        try {
            val url = URL("http://${_towerIp.value}:$TOWER_PORT/tower/healthcheck")

            val connection = url.openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.connect()

            connection.responseCode == HttpURLConnection.HTTP_OK
        } catch (e: Exception) {
            Log.e("SmsViewModel", "Error checking tower online", e)

            false
        }
    }

    fun updateUsername(text: String) {
        _username.value = text
    }

    fun updateMessageText(text: String) {
        if (text.length <= 280) {
            _messageText.value = text
        }
    }

    fun sendMessage() {
        val message = Message(
            messageId = UUID.randomUUID().toString(),
            username = _username.value,
            content = _messageText.value,
            timestamp = System.currentTimeMillis().toString()
        )

        viewModelScope.launch {
            var success = false
            try {
                success = sendMessageToServer(message)
            } catch (e: Exception) {
                Log.e("SmsViewModel", "Error sending message", e)
            }

            _sendError.value = !success
            if (success) {
                _messageText.value = ""
                refreshInbox()
            }
        }
    }

    private suspend fun sendMessageToServer(message: Message): Boolean = withContext(Dispatchers.IO) {
        try {
            val url = URL("http://${_towerIp.value}:$TOWER_PORT/send")

            val conn = url.openConnection() as HttpURLConnection
            conn.requestMethod = "POST"

            val jsonMessage = gson.toJson(message)

            conn.setRequestProperty("Content-Type", "application/json")
            conn.doOutput = true
            conn.outputStream.use { os ->
                os.write(jsonMessage.toByteArray())
            }

            Log.d("SmsViewModel", "Message (${message.content}) response code: ${conn.responseCode}")

            return@withContext conn.responseCode == HttpURLConnection.HTTP_OK
        } catch (e: Exception) {
            Log.e("SmsViewModel", "Failed to send message", e)

            return@withContext false
        }
    }

    fun refreshInbox() {
        viewModelScope.launch {
            var newMessages = emptyList<Message>()
            try {
                newMessages = fetchInbox()
            } catch (e: Exception) {
                Log.e("SmsViewModel", "Error refreshing inbox", e)
            }

            _messages.value = newMessages
        }
    }

    private suspend fun fetchInbox(): List<Message> = withContext(Dispatchers.IO) {
        try {
            val url = URL("http://${_towerIp.value}:$TOWER_PORT/inbox")

            val conn = url.openConnection() as HttpURLConnection
            conn.requestMethod = "GET"
            conn.connect()

            if (conn.responseCode != HttpURLConnection.HTTP_OK) {
                Log.e("SmsViewModel", "Failed to fetch inbox: ${conn.responseCode}")
                return@withContext emptyList()
            }

            val response = conn.inputStream.bufferedReader().use { it.readText() }

            Log.d("SmsViewModel", "Inbox response (${conn.responseCode}): $response")

            // Parse the response as a list of Message objects
            val type = object : com.google.gson.reflect.TypeToken<List<Message>>() {}.type
            return@withContext gson.fromJson<List<Message>>(response, type)
        } catch (e: Exception) {
            Log.e("SmsViewModel", "Error fetching inbox", e)

            return@withContext emptyList()
        }
    }
}

File: ios/Sada/ViewModels/SmsViewModel.swift
//
//  SmsViewModel.swift
//  Sada
//
//  Created by Nizar Mahmoud on 07/05/2025.
//

import Foundation
import Combine

private let towerPort = 8080
private let towerHealthcheck = "/healthcheck"
private let towerSend = "/send"
private let towerMessages = "/inbox"

@MainActor
class SmsViewModel: ObservableObject {
    @Published var deviceIp: String = ""
    @Published var towerIp: String = ""
    @Published var towerOnline: Bool = false
    @Published var sendError: Bool = false
    @Published var messageText: String = ""
    @Published var username: String = ""
    @Published var messages: [Message] = []

    let permissions: [String] = []
    let permissionsGranted: Bool = true

    init() {
        Task {
            self.deviceIp = NetworkUtils.getLocalIpAddress() ?? ""
            self.towerIp = NetworkUtils.getGatewayIpAddress() ?? ""
            self.towerOnline = await isTowerOnline()

            if self.towerOnline {
                await refreshMessages()
            }
        }
    }

    func updateUsername(_ text: String) {
        username = text
    }

    func updateMessageText(_ text: String) {
        if text.count <= 280 {
            messageText = text
        }
    }

    func sendMessage() {
        let message = Message(
            messageId: UUID().uuidString.lowercased(),
            username: username,
            content: messageText,
            timestamp: String(Int64(Date().timeIntervalSince1970 * 1000))
        )

        Task {
            let success = await sendMessageToServer(message: message)
            sendError = !success
            if success {
                messageText = ""
                await refreshMessages()
            }
        }
    }

    private func sendMessageToServer(message: Message) async -> Bool {
        guard !towerIp.isEmpty else { return false }
        let urlString = "http://\(towerIp):\(towerPort)\(towerSend)"
        guard let url = URL(string: urlString) else { return false }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 10

        do {
            let data = try JSONEncoder().encode(message)
            request.httpBody = data

            let (_, response) = try await URLSession.shared.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse else { return false }

            if httpResponse.statusCode == 200 {
                return true
            } else {
                print("SmsViewModel: Server returned status code:", httpResponse.statusCode)
                return false
            }
        } catch {
            print("SmsViewModel: Failed to send message:", error)
            return false
        }
    }

    private func isTowerOnline() async -> Bool {
        guard !towerIp.isEmpty else { return false }
        let urlString = "http://\(towerIp):\(towerPort)\(towerHealthcheck)"
        guard let url = URL(string: urlString) else { return false }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 5

        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse else { return false }
            return httpResponse.statusCode == 200
        } catch {
            print("SmsViewModel: Error checking tower online:", error)
            return false
        }
    }

    func refreshMessages() async {
        do {
            messages = try await fetchMessages()
        } catch {
            print("SmsViewModel: Error refreshing inbox:", error)
            messages = []
        }
    }

    private func fetchMessages() async throws -> [Message] {
        guard !towerIp.isEmpty else { return [] }
        let urlString = "http://\(towerIp):\(towerPort)\(towerMessages)"
        guard let url = URL(string: urlString) else { return [] }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 10

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw NSError(domain: "SmsViewModel", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid response"])
        }

        guard httpResponse.statusCode == 200 else {
            throw NSError(domain: "SmsViewModel", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: "Server returned status code: \(httpResponse.statusCode)"])
        }

        return try JSONDecoder().decode([Message].self, from: data)
    }
}
